Mathlib/LinearAlgebra/SpecialLinearGroup.lean

```diff
@@ -0,0 +1,605 @@
+    LinearMap.det (u : V →ₗ[R] V) = 1 := by
+  simp [← LinearEquiv.coe_det, u.prop]
+
+instance : CoeFun (SpecialLinearGroup R V) (fun _ ↦ V → V) where
+  coe u x := u.val x
+
+theorem ext_iff (u v : SpecialLinearGroup R V) : u = v ↔ ∀ x : V, u x = v x := by
+  simp only [← Subtype.coe_inj, LinearEquiv.ext_iff]
+
+@[ext]
+theorem ext (u v : SpecialLinearGroup R V) : (∀ x, u x = v x) → u = v :=
+  (SpecialLinearGroup.ext_iff u v).mpr
+
+section rankOne
+
+variable [Nontrivial R] [Module.Free R V] [Module.Finite R V]
+
+-- Move to other file ?
+theorem _root_.exists_linearEquiv_of_finrank_eq_one (d1 : Module.finrank R V = 1) :
+    Nonempty (R ≃ₗ[R] V) := by
+  let ⟨ι, b⟩ := (Module.Free.exists_basis R V).some
+  have : Finite ι := Module.Finite.finite_basis b
+  have : Fintype ι := Fintype.ofFinite ι
+  have : DecidableEq ι := Classical.typeDecidableEq ι
+  rw [Module.finrank_eq_card_basis b, Fintype.card_eq_one_iff] at d1
+  obtain ⟨i, hi⟩ := d1
+  exact ⟨{
+    toFun r := r • (b i)
+    invFun v := b.repr v i
+    left_inv r := by simp
+    right_inv v := b.repr.injective <| by ext j; simp [hi j]
+    map_add' r s := by simp [add_smul]
+    map_smul' r s := by simp [mul_smul] }⟩
+
+theorem exists_unique_eq_smul_id_of_finrank_eq_one (d1 : Module.finrank R V = 1) (u : V →ₗ[R] V) :
+    ∃! c : R, u = c • LinearMap.id := by
+  let e := (exists_linearEquiv_of_finrank_eq_one d1).some
+  set c := e.symm (u (e 1)) with hc
+  suffices u = c • LinearMap.id by
+    use c
+    simp only [this, true_and]
+    intro d hcd
+    rw [LinearMap.ext_iff] at hcd
+    specialize hcd (e 1)
+    simp only [LinearMap.smul_apply, LinearMap.id_coe, id_eq] at hcd
+    have hcd' := LinearEquiv.congr_arg (e := e.symm) hcd
+    simp only [map_smul, LinearEquiv.symm_apply_apply, smul_eq_mul, mul_one] at hcd'
+    rw [hcd']
+  ext x
+  have (x : V) : x = (e.symm x) • (e 1) := by simp [← LinearEquiv.map_smul]
+  rw [this x]
+  simp only [hc, map_smul, LinearMap.smul_apply, LinearMap.id_coe, id_eq]
+  rw [← this]
+
+/-- Endomorphisms of a free module of rank one are homotheties. -/
+noncomputable def LinearEquiv.smul_id_of_rank_one (d1 : Module.finrank R V = 1) :
+    R ≃ₗ[R] (V →ₗ[R] V) where
+  toFun := fun c ↦ c • LinearMap.id
+  map_add' c d := by ext; simp [add_smul]
+  map_smul' c d := by ext; simp [mul_smul]
+  invFun u := (exists_unique_eq_smul_id_of_finrank_eq_one d1 u).choose
+  left_inv c := by
+    simp [← (exists_unique_eq_smul_id_of_finrank_eq_one d1 (c • LinearMap.id)).choose_spec.2 c]
+  right_inv u :=
+    ((exists_unique_eq_smul_id_of_finrank_eq_one d1 u).choose_spec.1).symm
+
+theorem subsingleton_of_subsingleton (d1 : Module.finrank R V = 1) :
+    Subsingleton (SpecialLinearGroup R V) where
+  allEq u v := by
+    ext x
+    by_cases hx : x = 0
+    · simp [hx]
+    suffices ∀ (u : SpecialLinearGroup R V), (u : V →ₗ[R] V) = LinearMap.id by
+      simp only [LinearMap.ext_iff, LinearEquiv.coe_coe, LinearMap.id_coe, id_eq] at this
+      simp
```

Comment:
Squeeze

Reply:
I presumed you mean join the simp line with the next rw?